Automattic/woocommerce-payments
# Supply correct payment method instance to process_redirect_payment

Fixes #5854, Fixes #5781 

#### Changes proposed in this Pull Request

The crux of the the problem here is an [early return in `process_redirect_payment`](https://github.com/Automattic/woocommerce-payments/blob/9fa0bbbe007c1072776f13b401aeb530a4403f37/includes/payment-methods/class-upe-payment-gateway.php#L768). It's this early return that prevents the UPE payment intent from being removed from the session, causing the issue described in #5781. `UPE_Split_Payment_Gateway` extends `UPE_Payment_Gateway` and overrides `get_selected_payment_method` to redefine the logic with a simple check against `$this->stripe_id`. The problem is that `$this->stripe_id` is always `card` due to hooks in the parent class resetting the stripe id for it's own uses. So redirect payment methods such as `giropay` are always compared to `card` and we return `false`. That's fine as that's what `UPE_Payment_Gateway` needs and `UPE_Split_Payment_Gateway` doesn't need the `stripe_id` at this point in the execution. We really only need an instance of the payment method which we can get from the `upe_payment_method_map` via `WC_Payments::get_payment_method_by_id()`. 

### Testing

The goal is to test that `process_redirect_payment` fully completes. This would mean that [`remove_upe_intent_from_session()`](https://github.com/Automattic/woocommerce-payments/blob/9fa0bbbe007c1072776f13b401aeb530a4403f37/includes/payment-methods/class-upe-payment-gateway.php#L792) runs as well, which covers testing for both #5854 and #5781. Testing with Xdebug breakpoints has been problematic because the hooks that initiate this process fire multiple times, making it confusing as to whats actually occurring. I've found it best to just use `Logger::log()` to ensure the process completes.

1. Add a log statement immediately after [this line](https://github.com/Automattic/woocommerce-payments/blob/9fa0bbbe007c1072776f13b401aeb530a4403f37/includes/payment-methods/class-upe-payment-gateway.php#L829) indicating that the process has completed, e.g. `Logger::log( 'process_redirect_payment has completed' )`
2. Enable Split UPE Checkout via WCPay Dev Tools.
3. Go to the client DB and delete all rows in the `wp_woocommerce_sessions table`. This will ensure no `upe_payment_intent` is present yet.
4. Complete a checkout using a redirect payment method such as Giropay, making sure to **Authorize Test Payment** at Stripe.
5. Repeat step 4 at least 2 more times, using the same product and payment method. Reports have indicated that the payment intent error in #5781 requires 3 purchases.
6. Check logs for the message used in step 1. 
7. Ensure that the following error doesn't occur on the checkout page: 

>Error: The parameter application_fee_amount cannot be updated on a PaymentIntent after a capture has already been made.



-------------------


**Post merge**
